dev-python/*: bump twisted and some packages that depend on it to python 3.10

vaartis opened the pull request:
Separated from #20828, and will be rebased after it's merged. Waiting on the twisted keyword for ~alpha (https://bugs.gentoo.org/773451)

mgorny commented:
Did you manage to test PySocks? This bleeping thing technically requires ancient Tornado that can't work with py3.10 but you might be able to get them mostly working by collecting random patches from the Internet. I've basically started pinging people to stop using it as it's abandoned.

mgorny commented:
(I'm entirely happy if you managed to test it locally/in venv and not bother with making the tests work inside the ebuild)

vaartis commented:
> Did you manage to test PySocks? This bleeping thing technically requires ancient Tornado that can't work with py3.10 but you might be able to get them mostly working by collecting random patches from the Internet. I've basically started pinging people to stop using it as it's abandoned.

Tested it locally with requests on both 3.8 and 3.10, both seem to work fine.

vaartis commented:
Was able to get the tests to actually run, they work with 3.8 fine, in 3.10 tornado needs a single patch to run successfully and then the tests pass also.

mgorny commented:
> Was able to get the tests to actually run locally, they work with 3.8 fine, in 3.10 tornado needs a single patch to run successfully and then the tests pass also.

How much effort did you need to put into it? I'm wondering if there's a point in trying to make them ebuild-friendly after all.

vaartis commented:
> > Was able to get the tests to actually run locally, they work with 3.8 fine, in 3.10 tornado needs a single patch to run successfully and then the tests pass also.
> 
> How much effort did you need to put into it? I'm wondering if there's a point in trying to make them ebuild-friendly after all.

Changed the version in requirements_dev.txt for test server to 0.0.29 and after tox fetched the dependencies changed a single line in tornado from collections.MutableMapping to collections.abc.MutableMapping. That was enought for PySocks tests to pass.

mgorny commented:
BTW I'm surprised that Twisted works with py3.10. Back when I was doing py3.9 (and it was after py3.9 final), I had to fix a lot of issues upstream.

vaartis commented:
> BTW I'm surprised that Twisted works with py3.10. Back when I was doing py3.9 (and it was after py3.9 final), I had to fix a lot of issues upstream.

Surprisingly enough, yes, it seems to work fine, except those tests that depend on warning count. Perhaps thanks to your work on 3.9, but also may be because it's a relatively recent release (just two months ago).

mgorny commented:
Good news! Jinja has just been released in Python 3.10-compatible version, and I've pushed it to ::gentoo. Please remove jinja from the PR when rebasing ;-).

vaartis commented:
> Good news! Jinja has just been released in Python 3.10-compatible version, and I've pushed it to ::gentoo. Please remove jinja from the PR when rebasing ;-).

Great. Rebased and removed it, that's one problem less.

mgorny commented:
FYI, I've reverted asgiref because the warnings from the code break revdeps. Normally I'd say it's revdeps' fault for relying on warning-free dependencies but fighting 'fatal warnings' all over the place sounds like fighting windmills, so it's probably better to fix the issue in asgiref.

mgorny commented:
That said, I've forced keywords on twisted, so you may want to try rebasing and seeing where we can go from there.

vaartis commented:
Will look into it later today, probably going to also look into asigref stuff and patch whatever problems they have in their code and suggesting the patch upstream, depending on how much work it is.

vaartis commented:
https://github.com/django/asgiref/pull/262

Gathered a patch for asgiref to fix all their deprecation warnings, let's see what upstream says about them.

I saw in the bugzilla issue for alpha keywording that you have chose to instead mask 3.10 on alpha. Should I revbump all the ebuilds affected removing the keyword?

mgorny commented:
Don't remove keywords. The mask makes them irrelevant.

vaartis commented:
Now that upstream has approved the asgiref patches, I've re-added the ebuild here.

mgorny commented:
Great, thanks! Lemme check if/how much I can merge now.

vaartis commented:
There's still some keyword problem with semantic_version (nonsolvable deps), i think it's because of pypy3, I'll have to test that one and add it to the packages required.

mgorny commented:
I've merged the subset of packages that didn't depend on twisted. Could you rebase now?

dnspython-2 still fails for me.

vaartis commented:
Rebased. Can revbump + drop pypy3 from semantic_version? I don't know how it got there, but it needs a lot of depdendencies that haven't had pypy3 added yet, like all the packages from the last PR.

mgorny commented:
Sure, if no revdeps are broken. You can try removing it from all existing versions and using `check-revdep` to check if anything got broken (from mgorny-dev-tools).

vaartis commented:
> Sure, if no revdeps are broken. You can try removing it from all existing versions and using `check-revdep` to check if anything got broken (from mgorny-dev-tools).

According to the tool, there's dev-python/flask-login. If I remove it from both, then there are no more problems reported by it, neither with dev-python/semantic_version, nor with dev-python/flask-login.

vaartis commented:
Dropped pypy3 from both for now. Will look into the dnspython problem again tomorrow, probably. @mgorny, could you perhaps describe how you're testing it and it what environtment (maybe emerge --info?)

mgorny commented on dev-python/semantic_version/semantic_version-2.8.5-r1.ebuild:
```diff
@@ -3,7 +3,7 @@
 
```
Good news! I've figured out that my failures were due to pytest-django. Will commit a fix.

vaartis replied:
Great

mgorny commented on www-servers/tornado/tornado-6.1.ebuild:
```diff
@@ -32,6 +32,14 @@ distutils_enable_sphinx docs \
 
+python_prepare_all() {
+	# Disable deprecation-warnings-as-errors because tornado has a lot of stuff deprecated in 3.10
+	sed 's/warnings.filterwarnings("error", category=DeprecationWarning, module=r"tornado\\..*")//' \
```
This isn't going to fly. The warnings come from tornado itself, not the tests. You're shifting the problem to another package that will detect the warnings and fail.

vaartis replied:
I'll look into fixing it in the package itself and submitting it upstream

vaartis replied:
Seems like this one is too complicated for me, unfortunately, there are _a lot_ of deprecated things used.

vaartis replied:
Seems like this one's too hard for me, there's too much deprecated in there that I don't understand. I'll try at least opening an issue upstream about this.

mgorny replied:
Could you make a PR with at least these that you know how to fix? I think I've seen TLS-related stuff somewhere (urllib3? requests?), so you may want to try looking for a solution there.

vaartis replied:
When i try to fix things unfortunately it breaks even more :/ 
Unfortunately this one might be fit more for a real python developer

vaartis replied:
Got a response on the issue, they said it's probably going to be a major pain for them because of the get_current_loop deprecation.

mgorny commented on dev-python/twisted/twisted-21.2.0.ebuild:
```diff
@@ -4,7 +4,7 @@
 
 DISTUTILS_USE_SETUPTOOLS=rdepend
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
```
Does twisted pass for you?

I get:
```
  NewFetchTests
    testFetchAll ...                                                    [ERROR]
    testFetchAllUID ...                                                 [ERROR]
                                                [ERROR]
    testFetchBody ...                                                   [ERROR]
                                                  [ERROR]
    testFetchBodyParts ...                                              [ERROR]
                                             [ERROR]
    testFetchBodyStructureUID ...                                       [ERROR]
                                      [ERROR]
    testFetchBodyUID ...                                                [ERROR]
                                               [ERROR]
    testFetchEnvelope ... 
```
and at this point it hangs.

vaartis replied:
Strange. Yes, these do pass for me.

mgorny replied:
Yes.

vaartis commented on dev-python/urllib3/urllib3-1.26.4.ebuild:
```diff
@@ -64,6 +64,35 @@ python_test() {
 	local deselect=(
 		# TODO?
 		test/with_dummyserver/test_socketlevel.py::TestSocketClosing::test_timeout_errors_cause_retries
+		# Failed: Timeout >60.0s
```
Correct.

mgorny commented on dev-python/semantic_version/semantic_version-2.8.5-r1.ebuild:
```diff
@@ -3,7 +3,7 @@
 EAPI=7
 
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
```
You've lost the django dep here. But then, this makes it possible to merge it as-is for the time being, and add django when it's 3.10-ready. I'll file a bug not to forget about it.

vaartis replied:
Could you link the bug here?

mgorny replied:
Sure. https://bugs.gentoo.org/791367

mgorny commented:
I'm sorry, I was wrong. I've just retested and the twisted errors/hang is happening on 3.7 too. I'm going to investigate.

mgorny commented:
```
===============================================================================                                                                   [ERROR]                                                                                                                                           
Traceback (most recent call last):                                                                                                                
  File "/tmp/portage/dev-python/twisted-21.2.0/work/twisted-twisted-21.2.0-python3_7/test/lib/twisted/internet/test/test_gireactor.py", line 100, 
in test_gtkApplicationActivate                                                                                                                    
    reactor = gtk3reactor.Gtk3Reactor()                                                                                                           
  File "/tmp/portage/dev-python/twisted-21.2.0/work/twisted-twisted-21.2.0-python3_7/test/lib/twisted/internet/gtk3reactor.py", line 35, in __init
__                                                                                                                                                
    gireactor.GIReactor.__init__(self, useGtk=True)                                                                                               
  File "/tmp/portage/dev-python/twisted-21.2.0/work/twisted-twisted-21.2.0-python3_7/test/lib/twisted/internet/gireactor.py", line 73, in __init__
    _glibbase.GlibReactorBase.__init__(self, GLib, _gtk, useGtk=useGtk)                                                                             File "/tmp/portage/dev-python/twisted-21.2.0/work/twisted-twisted-21.2.0-python3_7/test/lib/twisted/internet/_glibbase.py", line 115, in __init_
_                                                                                                                                                     self._pending = self._gtk.events_pending                                                                                                        File "/usr/lib/python3.7/site-packages/gi/overrides/__init__.py", line 32, in __getattr__                                                       
    return getattr(self._introspection_module, name)                                                                                              
  File "/usr/lib/python3.7/site-packages/gi/module.py", line 124, in __getattr__                         
    self.__name__, name))                                                                                                                         
builtins.AttributeError: 'gi.repository.Gtk' object has no attribute 'events_pending'                                

twisted.internet.test.test_gireactor.GApplicationRegistrationTests.test_gtkApplicationActivate
```

What GTK+ version do you have?

mgorny commented:
Ok, I'm 90% convinced that it's GTK+4 getting in the way. I've unmerged it and I'm trying again. If all tests pass this time, I'm going to merge it.

mgorny commented:
No such luck:

```
  RunnersTests                                                                                                                           [35/1864]
    test_basicTrialIntegration ...                                         [OK]
    test_expectedResults ... /tmp/portage/dev-python/twisted-21.2.0/temp/environment: line 2913:   127 Segmentation fault      (core dumped) "${EPYTHON}" -m twisted.trial twisted
```

vaartis commented:
> Ok, I'm 90% convinced that it's GTK+4 getting in the way. I've unmerged it and I'm trying again. If all tests pass this time, I'm going to merge it.

That's curious. I don't have GTK installed at all.

mgorny commented:
I've filed https://twistedmatrix.com/trac/ticket/10200 for GTK+4, and I'll submit a patch to force GTK+3 shortly. If I don't forget, I'm going to test git twisted against py3.10. Or you can try (uninstall gtk:4 first):

```
git clone https://github.com/twisted/twisted
cd twisted
python3.10 setup.py build
PYTHONPATH=build/lib python3.10 -m twisted.trial twisted
```

vaartis commented:
> I've filed https://twistedmatrix.com/trac/ticket/10200 for GTK+4, and I'll submit a patch to force GTK+3 shortly. If I don't forget, I'm going to test git twisted against py3.10. Or you can try (uninstall gtk:4 first):

Thank you. I have tested the git twisted version and the test suite seems to pass fine, without GTK 4 that is. That said, I only just found out that there are two GTK packages in different categories. Turns out I did have GTK3 installed, so the tests do indeed work fine with it.

vaartis commented:
I've applied your patch to force gtk 3 to the newest version of twisted, and also tried to apply it to an older one, however it seems that older versions' tests no longer pass at all, because dev-python/pyopenssl is too new and we only have 20.0.1 in tree.

mgorny commented:
So you can't reproduce the segfault?

vaartis commented:
> So you can't reproduce the segfault?

Don't think so. It doesn't segfault for me.

mgorny commented:
So I've given it more thought and I've decided to merge it and stop blocking other packages on tornado. Worst case, other packages will hit the warnings and we'll either decide to wait or to ignore them there as well, and so on.

That is, 'it' without dnspython since that still fails for me and doesn't seem to be needed for the rest.

vaartis commented:
> So I've given it more thought and I've decided to merge it and stop blocking other packages on tornado. Worst case, other packages will hit the warnings and we'll either decide to wait or to ignore them there as well, and so on.
> 
> That is, 'it' without dnspython since that still fails for me and doesn't seem to be needed for the rest.

I see. Would be good to figure out what the problem with dnspython is..

mgorny commented:
I think it's related to output capture. I can get the test to pass by passing `-s` to pytest.

Could you try different versions of pytest, to see if this is some kind of regression?

vaartis commented:
> I think it's related to output capture. I can get the test to pass by passing `-s` to pytest.
> 
> Could you try different versions of pytest, to see if this is some kind of regression?

The only version of pytest that works with 3.10 is the latest one, others are broken on 3.10. It works on 8, 9 and 10 for me. The older versions of pytest work with 8 and 9 for me too.
